### PR TITLE
[EEG] EEG Acquisition form

### PR DESCRIPTION
**Pull request by regisoc**
## Brief summary of changes

EEG Acquisition form instrument

Screenshots:
<details>
  <summary>Click me</summary>

</details>

### CODE REVIEW

File: SQL/New_patches/2023-03-14_EEG_Acquisition_instrument.sql
Comment (driusan) [inline SQL/New_patches/2023-03-14_EEG_Acquisition_instrument.sql:1]: I don't think this belongs here. It's a sample RB instrument. @ridz1208 how do we usually handle those?

Also it would be a good idea to make it a json data instrument so that no table is required.

File: raisinbread/instruments/NDB_BVL_Instrument_EEG_Acquisition.class.inc
Comment (driusan) [inline raisinbread/instruments/NDB_BVL_Instrument_EEG_Acquisition.class.inc:1]: I don't see why this isn't a LINST instrument. It doesn't seem to do anything other than add basic field types and use XIN rules, both of which are supported by LINST.

**Review by christinerogers — COMMENTED** [review]
my first review comment:  the instrument should be called "EEG acquisition form" (it's missing the 3rd word) 
Agree with dave that instruments don't get table commits - see [other sample LORIS instruments](https://github.com/aces/Loris/tree/main/raisinbread/instruments)
Could you outline why this is PHP rather than LINST? is there typically branching logic, e.g.

File: raisinbread/instruments/NDB_BVL_Instrument_EEG_Acquisition.class.inc
Comment (regisoc) [reply raisinbread/instruments/NDB_BVL_Instrument_EEG_Acquisition.class.inc:1]: Agreed.

File: SQL/New_patches/2023-03-14_EEG_Acquisition_instrument.sql
Comment (regisoc) [reply SQL/New_patches/2023-03-14_EEG_Acquisition_instrument.sql:1]: If it removes the need for a table, sure! How can I do a JSON data instrument?

**Comment by regisoc** [comment]
Here are the screenshots. Did you test the form last Tuesday on my machine?

timepoint_view

**Comment by regisoc** [comment]
@christinerogers has been changed to `linst` + rules files + json data storage

**Comment by regisoc** [comment]
@christinerogers Also, do you want this EEG Acquisition form to be permanently in raisinbread?

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (ridz1208) [inline raisinbread/instruments/EEG_Acquisition_form.linst:2]: since those are in the .meta file they don't need to be in here

**Review by ridz1208 — CHANGES_REQUESTED** [review]
see comments 
+ there should not be new lines at the end on these files they can break the parser
+ there should probably be RB data that goes with it but that can come in a separate PR (at least the insert statemnts)

**Comment by ridz1208** [comment]
@GeorgeMurad can you make sure to test the data dict builder with it as well as testing the instrument itself

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (regisoc) [reply raisinbread/instruments/EEG_Acquisition_form.linst]: Lines removed

File: raisinbread/instruments/EEG_Acquisition_form.meta
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.meta:2]:
```suggestion
table{@}eeg_acquisition_form
```
see how mri parameter form table vs testname :  https://github.com/aces/Loris/blob/main/raisinbread/RB_files/RB_mri_parameter_form.sql

File: raisinbread/instruments/EEG_Acquisition_form.meta
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.meta:1]:
```suggestion
testname{@}eeg_acquisition_form
```
see: https://github.com/aces/Loris/blob/main/raisinbread/instruments/NDB_BVL_Instrument_mri_parameter_form.class.inc#L54

File: raisinbread/instruments/EEG_Acquisition_form.rules
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.rules:1]: update these field names below after changes above...

File: raisinbread/instruments/EEG_Acquisition_form.rules
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.rules:4]: This `u` is inconsistent --- Please grep aces/Loris for "ehavio" and use the most frequent spelling you find -- it's either behavior* or behaviour*
regis: it's a canadian/us spelling issue - that's why i'm recommending you go with the most frequent spelling loris uses.

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:51]: remove this comment.  the "video" reference is too project-specific

Comment (christinerogers) [reply raisinbread/instruments/EEG_Acquisition_form.linst:51]: please remove this/similar video comment in all other places

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:49]: does refusal make sense for an end time field?

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:65]:
```suggestion
textarea{@}vep_eeg_mood_rating_comments{@}
Comments on subject mood:
```
please change this question for all tasks

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:66]:
```suggestion
textarea{@}vep_comments{@}Comment on the VEP task:
```
Change this on all tasks

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:2]:
```suggestion
header{@}{@}EEG Acquisition Form
```
capital F

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/EEG_Acquisition_form.linst:1]:
```suggestion
title{@}EEG Acquisition Form
```

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (christinerogers) [reply raisinbread/instruments/EEG_Acquisition_form.linst:2]: Can this line be removed or Is this needed?  
See your screenshots - there's an extra "EEG Acquisition" header line in the form above General Informations which should be removed.

**Review by christinerogers — CHANGES_REQUESTED** [review]
Hi @regisoc 
Great first draft,   Once you get a chance to make these changes (shouldn't take more than an hour) and let me know in slack when it's ready for review.  
 When you commit, please take it out of Draft
 
Tip from the Loris meeting discussion: how to check for `behavio(u)r` spelling consensus, and put your findings back in the agenda when done : 
 * check the likeliest front-end modules and get a sense from reading them: - do they use vior or viour ?  this is the one we want to be consistent with. 
* For the whole codebase You can find out the frequency count for behavio(u)ral in the corpus by grepping and counting recursively on your whole local codebase something like  `grep -ro "ehavio." $yourLorisDirectory | sort | uniq -c 1`

**Comment by regisoc** [comment]
Excluding `node_modules` and `project` folder to target only LORIS Core code, it seems the big winner is `behaviour`.
```bash
# behavior
/var/www/Loris$ grep -ro --exclude-dir=project --exclude-dir=node_modules "ehavior" . | sort | uniq -c | awk '{s+=$1} END {print s}'
3598
# behaviour
/var/www/Loris$ grep -ro --exclude-dir=project --exclude-dir=node_modules "ehaviour" . | sort | uniq -c | awk '{s+=$1} END {print s}'
5067
```

File: raisinbread/instruments/EEG_Acquisition_form.linst
Comment (regisoc) [reply raisinbread/instruments/EEG_Acquisition_form.linst:49]: Good question, it was generated through the parser. Looking at the first iteration, it was this line: 
`$this->addHourMinElement("eeg_session_start_time", "EEG session Start time");`
but the code automatically generated two grouped fields (`eeg_session_start_time` and `eeg_session_start_time_status`). The `eeg_session_start_time_status` field has these values.
The code is here: https://github.com/aces/Loris/blob/main/php/libraries/NDB_BVL_Instrument.class.inc#L1741
I will remove both `*_status` fields for this case.

File: raisinbread/instruments/EEG_Acquisition_form.rules
Comment (regisoc) [reply raisinbread/instruments/EEG_Acquisition_form.rules:2]: Not sure of this one, what changes are you refering to?

File: raisinbread/instruments/EEG_Acquisition_form.rules
Comment (regisoc) [reply raisinbread/instruments/EEG_Acquisition_form.rules]: `behaviour` it is, see comment below

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:16]: Please add an option : "No measurement taken" -- it's really only for pediatric studies that they measure.
please add a rule:  if "no measurement taken" is selected, the following field (for the number) is optional.

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:20]:
```suggestion
select{@}subject_away_60cm_from_monitor{@}Subject is 60cm away from monitor?{@}'yes'=>'Yes'{-}'no'=>'No'
```
add a "don't know" option -- a lot of projects don't collect this.

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:23]:
```suggestion
text{@}device_id{@}Device ID:
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:24]:
```suggestion
select{@}impedance_measured_before_applied{@}Impedances measure before application on subject?{@}'yes'=>'Yes'{-}'no'=>'No'
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:26]:
```suggestion
select{@}bad_channels_measured_after_applied{@}Bad channels measured after application?{@}'yes'=>'Yes'{-}'no'=>'No'
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:28]:
```suggestion
header{@}{@}<h6>About placement:<br/> <b>Excellent</b> - meets all criteria for proper placement.<br/> <b>Average</b> - 1-2cm shift from proper placement (1 pinky finger width distance).<br/> <b>Poor</b> - >2 cm shift from proper placement (>1 pinky finger width distance). </h6>
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:29]:
```suggestion
select{@}placement_rating{@}Rate placement:{@}'na'=>'n/a'{-}'excellent'=>'Excellent'{-}'average'=>'Average'{-}'poor'=>'Poor'
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:30]:
```suggestion
textarea{@}placement_comment{@}Comments on placement
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:31]:
```suggestion
select{@}placement_photos_taken{@}Placement photos taken?{@}'yes'=>'Yes'{-}'no'=>'No'
```

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:43]: @regisoc can you remind me why these questions are located down here rather than with the questions on impedance testing?

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:62]:
```suggestion
select{@}task1_subject_mood_rating{@}Task 1 EEG Mood Rating{@}NULL=>''{-}'1'=>'1 - Screaming, fussed out'{-}'2'=>'2'{-}'3'=>'3'{-}'4'=>'4'{-}'5'=>'5 - Very calm the whole time'
```
* please replace "eeg" with "subject" for all similar "mood" questions including comments questions 
* please add an option for "not known"

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:58]: What does "Task 1 used" mean ?  par rapport a "Task 1 eeg collected" -- 
Is it possible for the Task  to be used with no eeg acquired?   or for task eeg to be acquired without the Task to be "used" ?  What other word is more appropriate than "used"?

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.linst:21]: Add 1 more questions: 
```suggestion
textarea{@}subject_away_60cm_from_monitor_reason{@}If No, please explain the reason
select{@}head_rest_used{@}Was a head rest used?{@}'yes'=>'Yes'{-}'no'=>'No'{-}'dnk'=>'Don't Know'
```

**Review by christinerogers — CHANGES_REQUESTED** [review]
Hi @regisoc  - thanks for incorporating the previous changes requested. 
I've run the latest version by an external eeg expert. My comments and suggestions below reflect their suggestions which will make this form usable across many more eeg studies. 
These include: 
* generalizing the device language so that cap and montage studies (not just net) can use this form
* in general, let's add a "not known" to yes/no questions. 
* I added one question: was a head rest used? 
* for non-pediatric studies, tweaked some language 
* also changed the subject mood field name (it was eeg_mood) 
I can test this on your VM tomorrow, just let me know when updated and ready -- thanks

File: raisinbread/instruments/eeg_acquisition_form.linst
Comment (regisoc) [reply raisinbread/instruments/eeg_acquisition_form.linst:43]: I think it comes from the HBCD EEG acquisition form, we can put them up here to be more generic.

**Comment by regisoc** [comment]
@christinerogers made the changes! Ready for review.I will send you the steps for testing on my vm.

**Review by christinerogers — APPROVED** [review]

**Review by christinerogers — CHANGES_REQUESTED** [review]
Hi @regisoc I started re-reviewing on your vm --

the dropdown selects all need a null value -- i just noticed now they're not formed in the typical way. 
they should all have a first NULL option like this: 
https://github.com/aces/Loris/blob/main/raisinbread/instruments/bmi.linst#L7
See example BMI form on Demo.loris.ca
without this null first option, In your form all the selects automatic load the first option by default -- see example screenshot: 
Screen Shot 2023-03-31 at 5 39 43 PM

This should be a quick fix.  Let me know when it's ready and I'll re-review by Monday lunch.

**Comment by regisoc** [comment]
Done you can refresh the page you are on to see the changes

**Review by christinerogers — CHANGES_REQUESTED** [review]
I'm just now realizing you generated this linst manually instead of with the Instrument builder. 
Please load this in the instrument builder module (on your VM or demo), then save it, and *compare* the output -- you should see among other Linst format points : 
*   Null options are added to selects (per my previous review) 
* not_answered option for each dropdown
* time type is i think deprecated; that's why it doesn't show in your VM front-end 

Please also : 
* change do not know to not_known "not known" for each dropdown
* remove `:` at the end of a field label (because it looks random)
* see other gallicismes or inconsistencies i flagged in comments below

**Comment by christinerogers** [comment]
> I'm just now realizing you generated this linst manually instead of with the Instrument builder. 
> 
> Please load this in the instrument builder module (on your VM or demo), then save it, and *compare* the output -- you should see among other Linst format points : 
> 
> *   Null options are added to selects (per my previous review) 
> 
> * not_answered option for each dropdown
> 
> * time type is i think deprecated; that's why it doesn't show in your VM front-end 
> 
> 
> 
> Please also : 
> 
> * change do not know to not_known "not known" for each dropdown
> 
> * remove `:` at the end of a field label (because it looks random)
> 
> * see other gallicismes or inconsistencies i flagged in comments below

@regisoc could you confirm a few things : 
- does your latest push address my last comments ? 
- does the instrument work? Does it successfully save data in All fields in the back-end table on your VM?
- when there's a moment Monday, thanks for updating the screenshots in PR
Let me know when it's ready for final review, thanks --

**Comment by regisoc** [comment]
@christinerogers to confirm the few things: 
- yes, my latest push address your comments. I added one more because I reviewed it and saw one participant->subject rule missing.
- yes, the instrument works on my machine, I can see data are written in the database. It should be good to test on someone else machine just to be sure.
- it is ready for final review. I will update the screenshots right away.

**Review by christinerogers — APPROVED** [review]

**Review by christinerogers — CHANGES_REQUESTED** [review]
I see a few minor issues (#8491) which shouldn't hold up the 25 merge / release 

@regisoc please just add one missing rule, and i'll approve.  Hopefully this last change will take 90 seconds maximum: 

- [ ] for the question that follows `Subject 60m away` :  `if no, please explain` needs a rule (only required if the previous response is "no")

Thanks -- very nearly there

**Comment by christinerogers** [comment]
@GeorgeMurad could you please run this through the [data dict builder as @ridz1208 requested](https://github.com/aces/Loris/pull/8443#issuecomment-1477945828)

@ridz1208 your review (below) seems to be addressed -- I'll dismiss,  Want to re-review ? 

> * there should not be new lines at the end on these files they can break the parser. `-> see commit  #d7d44c0`
> * there should probably be RB data that goes with it but that can come in a separate PR (at least the insert statemnts)

 -> RB data can come for next release, not necessary now -- per last Loris meeting


otherwise this is one small rule change away from merge-ready imo,  @driusan

File: raisinbread/instruments/eeg_acquisition_form.rules
Comment (christinerogers) [inline raisinbread/instruments/eeg_acquisition_form.rules:9]:
```suggestion
subject_away_60cm_from_monitor_reason{-}Required if subject cannot be 60cm away from monitor{-}subject_away_60cm_from_monitor{@}=={@}no
```

Comment (christinerogers) [reply raisinbread/instruments/eeg_acquisition_form.rules:9]: @regisoc just spotted another typo to correct - thanks  (message appears in front end)

**Review by christinerogers — CHANGES_REQUESTED** [review]

**Review by christinerogers — APPROVED** [review]

**Comment by regisoc** [comment]
@driusan ready for a final review

**Pull request merged**
